nkms/storage/encrypted_file.py

```diff
@@ -1,114 +1,41 @@
         1) The ciphertext -- This is the chunked and encrypted ciphertext
         2) The header -- This contains the metadata of the ciphertext that
             tells us how to decrypt it, or add more data.
 
         :param bytes key: Symmetric key to use for encryption/decryption
-        :param string/bytes path: Path of file to open
-        :param string mode: Mode to use when opening file, default is 'rb'
+        :param bytes path: Path of ciphertext file to open
+        :param bytes header_path: Path of header file
         """
-        self.path = path
-        self.mode = mode
-
         cipher = symmetric_from_algorithm(default_algorithm)
         self.cipher = cipher(key)
 
-    def _build_header(self, version=100, nonce=None, keys=None,
-                      chunk_size=1000000, num_chunks=0, msg_len=0):
-        """
-        Builds a header and returns the msgpack encoded form of it.
-
-        :param int version: Version of the NuCypher header
-        :param bytes nonce: Nonce to write to header, default is random(20)
-        :param list keys: Keys to write to header
-        :param int chunk_size: Size of each chunk in bytes, default is 1MB
-        :param int num_chunks: Number of chunks in ciphertext, default is 0
-        :param int msg_len: Length of the encrypted ciphertext in total
+        # Opens the header file and parses it, if it exists. If not, creates it
+        self.header_path = header_path
+        self.header_obj = Header(self.header_path)
 
-        :return: (header_length, encoded_header)
-        :rtype: Tuple(int, bytes)
-        """
-        if not nonce:
-            nonce = random(20)
-
-        self.header = {
-            'version': version,
-            'nonce': nonce,
-            'keys': keys,
-            'chunk_size': chunk_size,
-            'num_chunks': num_chunks,
-            'msg_len': msg_len,
-        }
-
-        try:
-            encoded_header = msgpack.dumps(self.header)
-        except ValueError as e:
-            raise e
-        self.header_length = len(encoded_header)
-        return (self.header_length, encoded_header)
-
-    def _encode_header(self):
-        """
-        Returns a msgpack encoded header and the length of it in bytes ready to
-        be written to the file_obj.
-
-        :return: (encoded_header, header_length_bytes)
-        :rtype: Tuple(bytes, bytes)
-        """
-        header_length_bytes = self.header_length.to_bytes(4, byteorder='big')
-        encoded_header = msgpack.dumps(self.header)
-        return (encoded_header, header_length_bytes)
-
-    def _update_header(self, header):
-        """
-        Updates the self.header with the key/values in header, then updates
-        the header length.
+        self.path = path
 
-        :param dict header: Dict to update self.header with
+        # Always seek the beginning of the file on first open
+        self.file_obj = open(self.path, mode='a+b')
```

Comment:
When we open a file in the default mode `a+b`, we're doing a few things:

1. Creating the file if it doesn't exist.
2. Opening for append

I figured that if we're reading the data from it, then we'll need to be at the beginning of the file anyway. If the file is new, then it doesn't matter and we'll be at the very beginning.

If we're writing, then it seeks to the end (see L85) -- ` self.file_obj.seek(0, os.SEEK_END)`.

Honestly, I'm trying to build something that "just works" so I can work on the cryptographic API since that's our main priority right now.

I would love to change this and make this interface really nice and pretty, but I have other priorities at the moment.

I'd love to hear your solutions on making this better, and if you feel like you want to implement them, please feel free to do so. Admittedly, this is one area that I could very much so improve on and would love all sorts of help on it.

Reply:
Agree on making something which just works. If it's a fundamental architectural decision, we do it now. If it's nice-to-have, we can always improve later